### ISSUE TITLE
Add an option to add BEGIN and COMMIT around migrations when generating them for PostgreSQL

### ISSUE THREAD

**Issue by alecmev** [prisma/prisma#8080]
Sister issue of https://github.com/prisma/prisma/issues/7641, which is specific to SQL Server. Prompted by comments in https://github.com/prisma/prisma/discussions/3774. I believe that migrations running outside of a transaction must be an exception, not a rule, since guaranteeing 100% clean state (so the migration is all or nothing) is more important than higher performance or higher chance of success. Where it's really impractical to keep the migration in a transaction, e.g. when touching too many rows in a table that's too busy, the relevant statements can be put outside the transaction manually.

**Comment by tomhoule** [comment]
From internal discussions, we see a lot of advantages to having the option to wrap migrations in a transaction have the effect of manually adding a `BEGIN` and a `COMMIT` to generated migrations by default, rather than affecting how Migrate applies migrations directly: it's less complex to implement (applying migrations should be very predictable and deterministic), it's persisted in the migration itself, and it's more flexible, since users can opt in or out on a per-migration basis (or even for parts of a migration).

**Comment by janpio** [comment]
This could also be a parameter for commands that create a migration, `--wrap-in-transaction` or `--use-transaction` for example that would then cause these to be rendered in the migration SQL by default.

**Comment by casey-chow** [comment]
I'm not sure I understand what's going on here, because on one hand I'm seeing a request to put migrations inside transactions by default (implying that migrations don't run inside a transaction) but then this discussion question implies that they are being run: https://github.com/prisma/prisma/discussions/10601

What's the current behavior?

**Comment by mircohaug** [comment]
From [this blogpost](https://www.prisma.io/blog/prisma-migrate-dx-primitives#what-if-schema-migrations-were-atomic) it seems that the transaction wrapping behavior is database engine specific. 
 * SQL Server and Azure SQL: Migrate explicitly wraps the generated migration in a transaction.
 * PostgreSQL: You can opt-in by adding BEGIN; and COMMIT; to the generated schema migrations. By default, Migrate does not wrap migrations in a transaction.
 * MySQL: Transactional DDL is not supported.

**Comment by casey-chow** [comment]
Yeah I think that blog post is incorrect in light of the behavior I saw in #10601 (and have myself seen). It would be helpful to have this behavior in documentation, whereas in blog posts which are point-in0time.

**Comment by capaj** [comment]
also this "explicit" transaction mode would fix the problem when trying to create concurrent indexes currently:  
```
Database error:
ERROR: CREATE INDEX CONCURRENTLY cannot run inside a transaction block

DbError { severity: "ERROR", parsed_severity: Some(Error), code: SqlState(E25001), message: "CREATE INDEX CONCURRENTLY cannot run inside a transaction block", detail: None, hint: None, position: None, where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some("xact.c"), line: Some(3389), routine: Some("PreventInTransactionBlock") }

```

**Comment by korygorsky** [comment]
+1 Would love to see this.

**Comment by chriskuech** [comment]
Any update on at least explicitly documenting whether prisma applies migrations in a transaction?

**Comment by Joepena** [comment]
+1 on documenting if a transaction is being applied or not

**Comment by amj311** [comment]
+1 Transactionally safe migrations feels like a necessary feature before I would support full adoption of prisma in my professional projects

**Comment by chriskuech** [comment]
Just to document more weirdness around this subject--

PostgreSQL enum migrations _do_ wrap with `BEGIN`/`COMMIT`. Yet based on this thread, it seems redundant...

**Comment by b1rdex** [comment]
> based on this thread, it seems redundant

No, it's not redundant. Wrapping migrations in transactions (implicitly on execution) is implemented only for MSSQL

Here is actual code: https://github.com/search?q=repo%3Aprisma%2Fprisma-engines%20render_begin_transaction&type=code

So the summary in https://github.com/prisma/prisma/issues/8080#issuecomment-1245083538 is actual.

**Comment by aseemk** [comment]
I wrote this comment in [a discussion](https://github.com/prisma/prisma/discussions/3774#discussioncomment-8768285); copying here if this issue is a better place. (Apologies if not!)

+1 to all the requests asking for transactions by default. Just got bitten by this, and it really sucks to have to debug and fix a DB with a half-migrated schema.

The lack of this safe behavior is what results in the need for `prisma migrate diff` and `resolve` complexity. In 99% of cases, I just want my migrations to succeed all-or-nothing. And in the few times where I need to make a change that can't run in a transaction (like `CREATE INDEX CONCURRENTLY`), I'd rather wrap _that_ transaction in `COMMIT; ...; BEGIN;` rather than wrap _every other_ transaction in `BEGIN; ...; COMMIT;`

Explicit `BEGIN` and `COMMIT` statements in every _generated_ migration by default is also reasonable, and would be great to have as a bare minimum.

Thank you!

**Comment by tony-mack-dealstack** [comment]
```
fn render_begin_transaction(&self) -> Option<&'static str> {
    Some("BEGIN;")
}

fn render_commit_transaction(&self) -> Option<&'static str> {
    Some("COMMIT;")
}

```
in 
`schema-engine/connectors/sql-schema-connector/src/sql_renderer/postgres_renderer.rs`

would add the begin commit into the file generation like it does for MSSQL

then if we really wanted we could remove it instead for trying to remember to put it there?
